lib/kessel/auth.rb

```diff
@@ -0,0 +1,193 @@
+            'OAuth functionality requires the openid_connect gem. Add "gem \'openid_connect\'" to your Gemfile.'
+    end
+
+    # OpenID Connect Client Credentials flow implementation using discovery.
+    #
+    # This provides a secure OIDC Client Credentials flow implementation with
+    # automatic endpoint discovery. Works seamlessly OIDC-compliant providers
+    # that support discovery.
+    #
+    # @example
+    #   oauth = OAuth.new(
+    #     client_id: 'kessel-client',
+    #     client_secret: 'super-secret-key',
+    #     token_endpoint: 'https://my-domain/auth/realms/my-realm/protocol/openid-connect/token'
+    #   )
+    #
+    #   # Get current access token (automatically cached and refreshed)
+    #   token = oauth.token
+    class OAuth2ClientCredentials
+      include Kessel::Auth
+
+      # OAuth client identifier.
+      # @return [String] The client ID
+      attr_reader :client_id
+
+      # OIDC issuer URL for discovery.
+      # @return [String] The issuer URL
+      attr_reader :token_endpoint
+
+      # Creates a new OIDC client with specified token endpoint.
+      #
+      # @param client_id [String] OIDC client identifier
+      # @param client_secret [String] OIDC client secret
+      # @param token_endpoint [String] OIDC token endpoint URL
+      #
+      # @raise [OAuthDependencyError] if the openid_connect gem is not available
+      # @raise [OAuthAuthenticationError] if authentication fails
+      #
+      # @example
+      #   oauth = OAuth.new(
+      #     client_id: 'my-app',
+      #     client_secret: 'secret',
+      #     token_endpoint: 'https://my-domain/auth/realms/my-realm/protocol/openid-connect/token'
+      #   )
+      def initialize(client_id:, client_secret:, token_endpoint:)
+        check_dependencies!
+
+        @client_id = client_id
+        @client_secret = client_secret
+        @token_endpoint = token_endpoint
+        @token_mutex = Mutex.new
+      end
+
+      # Gets the current access token with automatic caching and refresh.
+      #
+      # Uses OIDC Client Credentials flow with automatic token caching,
+      # expiration checking, and refresh logic.
+      #
+      # @return [String] A valid access token
+      # @raise [OAuthAuthenticationError] if token acquisition fails
+      #
+      # @example
+      #   token = oauth.token
+      #   # Use token in Authorization header: "Bearer #{token}"
+      def get_token(force_refresh: false)
+        return @cached_token if !force_refresh && token_valid?
+
+        @token_mutex.synchronize do
+          @cached_token = nil if force_refresh
+
+          # Double-check: another thread might have refreshed the token
+          return @cached_token if token_valid?
+
+          @cached_token = refresh
+
+          return @cached_token
+        rescue StandardError => e
+          raise OAuthAuthenticationError, "Failed to obtain client credentials token: #{e.message}"
+        end
+      end
+
+      def refresh
```

Comment:
I thought the idea with the spec here was to only have 1 method for refreshing using `get_token` for both getting a cached token and refreshing if needed.